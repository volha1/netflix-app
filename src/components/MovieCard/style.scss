@import '../../common/scss/base.scss';

.card {
  display: flex;
  flex-direction: column;
  position: relative;
  margin-bottom: 50px;
  margin-right: 50px;
  cursor: pointer;

  &:nth-child(3n) {
    margin-right: 0px;
  }

  .movie-cover {
    position: relative;
    width: $card-width;
    height: 455px;

    .movie-img {
      width: $card-width;
      height: 455px;
      margin-bottom: 20px;
    }

    .menu-icon {
      position: absolute;
      right: 10px;
      top: 10px;
      visibility: hidden;
      opacity: 0;
      transition: opacity 0.3s linear;
    }
  }

  .title-wrapper {
    display: flex;
    justify-content: space-between;
    align-items: center;
    opacity: 0.7;
    margin-top: 10px;
    margin-bottom: 5px;
    width: 322px;

    .card-title {
      font-size: 18px;
    }

    .year {
      border: 1px solid $gray;
      border-radius: 5px;
      padding: 5px 15px;
    }
  }

  .genres {
    width: $card-width;
    font-size: 14px;
    opacity: 0.7;
  }

  &:hover > .movie-cover > .menu-icon {
    visibility: visible;
    opacity: 1;
  }
}
