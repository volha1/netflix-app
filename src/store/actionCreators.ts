import Movie from '../entity/Movie';
import types from './actionTypes';

const addAllMovies = (payload: Array<Movie>): { type: string; payload: Array<Movie> } => {
  return {
    type: types.ADD_ALL_MOVIES,
    payload,
  };
};

const markForDeletion = (payload) => ({ type: types.MARK_FOR_DELETION, payload });

const deleteMovie = () => ({ type: types.DELETE_MOVIE });

export { addAllMovies, markForDeletion, deleteMovie };
