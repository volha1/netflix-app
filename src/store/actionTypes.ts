const types = {
  ADD_ALL_MOVIES: 'ADD_ALL_MOVIES',
  MARK_FOR_DELETION: 'MARK_FOR_DELETION',
  DELETE_MOVIE: 'DELETE_MOVIE',
};

export default types;
