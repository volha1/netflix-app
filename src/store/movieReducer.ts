import types from './actionTypes';

const defaultState = {
  movies: [],
  movieIdForDeletion: '',
};

const movieReducer = (state = defaultState, action) => {
  switch (action.type) {
    case types.ADD_ALL_MOVIES:
      return { ...state, movies: [...action.payload] };
    case types.MARK_FOR_DELETION:
      return { ...state, movieIdForDeletion: action.payload };
    case types.DELETE_MOVIE:
      return { ...state, movies: state.movies.filter((movie) => movie.id !== state.movieIdForDeletion) };
    default:
      return state;
  }
};

export default movieReducer;
