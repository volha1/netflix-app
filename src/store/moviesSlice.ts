/* eslint-disable @typescript-eslint/no-explicit-any */
import { createSlice, createAsyncThunk, AnyAction } from '@reduxjs/toolkit';
import { url } from '../helpers/constants';
import Movie from '../types/Movie';

type MovieItem = {
  id?: any;
  title: any;
  vote_average: any;
  release_date?: any;
  poster_path: any;
  genres: any;
  overview: any;
  runtime: any;
};

const deleteMovieById = createAsyncThunk('movies/deleteById', async (id: string): Promise<void> => {
  const response = await fetch(`${url}/movies/${id}`, { method: 'DELETE' });

  if (!response.ok) {
    throw new Error(response.statusText);
  }
});

const getAllMoviesSorted = createAsyncThunk<Movie[], any>('movies/sortMovies', async (params): Promise<Movie[]> => {
  const requestParams = [];

  for (const [key, value] of Object.entries(params)) {
    if (value) {
      requestParams.push(`${key}=${value}`);
    }
  }

  const response = await fetch(`${url}/movies?${requestParams.join('&')}`);

  if (!response.ok) {
    throw new Error(response.statusText);
  }

  const json = await response?.json();
  const movies = json.data.map((movie: MovieItem) => {
    return {
      id: movie.id,
      title: movie.title,
      voteAverage: movie.vote_average,
      releaseDate: movie.release_date,
      imgPath: movie.poster_path,
      genres: movie.genres,
      overview: movie.overview,
      runtime: movie.runtime,
    };
  });

  return movies;
});

const createMovie = createAsyncThunk('movies/createMovie', async (movie: Movie): Promise<void> => {
  const movieItem: MovieItem = {
    title: movie.title,
    vote_average: Number(movie.voteAverage),
    poster_path: movie.imgPath,
    overview: movie.overview,
    runtime: Number(movie.runtime),
    genres: movie.genres,
  };

  if (movie.releaseDate) {
    movieItem.release_date = movie.releaseDate;
  }

  const response = await fetch(`${url}/movies`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(movieItem),
  });

  if (!response.ok) {
    throw new Error(response.statusText);
  }
});

const updateMovie = createAsyncThunk('movies/createMovie', async (movie: Movie): Promise<any> => {
  const movieItem: MovieItem = {
    id: movie.id,
    title: movie.title,
    vote_average: Number(movie.voteAverage),
    poster_path: movie.imgPath,
    overview: movie.overview,
    runtime: Number(movie.runtime),
    genres: movie.genres,
  };

  if (movie.releaseDate) {
    movieItem.release_date = movie.releaseDate;
  }

  const response = await fetch(`${url}/movies`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(movieItem),
  });

  if (!response.ok) {
    throw new Error(response.statusText);
  }
});

const isError = (action: AnyAction): boolean => {
  return action.type.endsWith('rejected');
};

const moviesSlice: any = createSlice({
  name: 'movies',
  initialState: {
    movies: [] as Movie[],
    movieIdForDeletion: '',
    loadingStatus: false,
    error: '',
    movieForEditing: {},
  },
  reducers: {
    markMovieForDeletion(state, action) {
      state.movieIdForDeletion = action.payload;
    },
    saveMovieForEditing(state, action) {
      state.movieForEditing = action.payload;
    },
    clearError(state, action) {
      state.error = action.payload;
    },
  },
  extraReducers: (builder) => {
    builder.addCase(getAllMoviesSorted.pending, (state) => {
      state.loadingStatus = true;
    });
    builder.addCase(getAllMoviesSorted.fulfilled, (state, action) => {
      state.loadingStatus = false;
      state.movies = action.payload;
    });
    builder.addCase(deleteMovieById.fulfilled, (state) => {
      state.movies = state.movies.filter((movie: Movie) => {
        return movie.id !== state.movieIdForDeletion;
      });
    });
    builder.addMatcher(isError, (state, action) => {
      state.loadingStatus = false;
      state.error = action.error.message || 'Server error';
    });
  },
});

const { markMovieForDeletion, saveMovieForEditing, clearError } = moviesSlice.actions;

export {
  markMovieForDeletion,
  saveMovieForEditing,
  deleteMovieById,
  getAllMoviesSorted,
  createMovie,
  updateMovie,
  clearError,
};
export default moviesSlice.reducer;
