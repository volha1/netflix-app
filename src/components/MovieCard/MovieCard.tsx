import React, { ReactElement, useState, memo, Dispatch, SetStateAction, useCallback } from 'react';
import './style.scss';
import Movie from '../../entity/Movie';
import menuIcon from '../../common/assets/svg/menu-icon.svg';
import Menu from '../Menu';
import { getYear } from '../../helpers/utils';

type MovieCardProps = {
  movie: Movie;
  deleteMovie: () => void;
  selectMovie: Dispatch<SetStateAction<Movie>>;
};

const MovieCard = ({ movie, deleteMovie, selectMovie }: MovieCardProps): ReactElement => {
  const [isMenuVisible, setMenuVisible] = useState(false);

  const handleMenuClick = useCallback(() => {
    setMenuVisible(!isMenuVisible);
  }, [isMenuVisible]);

  const handleCardClick = (): void => {
    selectMovie(movie);
  };

  return (
    <div className="card" onClick={handleCardClick}>
      <div className="movie-cover">
        <img className="movie-img" src={movie.imgPath} alt={movie.title} />
        <img className="menu-icon" src={menuIcon} alt="Menu" onClick={handleMenuClick} />
        <Menu visible={isMenuVisible} handleClose={handleMenuClick} deleteMovie={deleteMovie} />
      </div>
      <div className="title-wrapper">
        <p className="card-title">{movie.title}</p>
        <p className="year">{getYear(movie.year)}</p>
      </div>
      <p className="genres">{movie.genres.join(', ')}</p>
    </div>
  );
};

export default memo(MovieCard);
