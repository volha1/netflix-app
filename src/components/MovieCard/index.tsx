import React, { ReactElement } from 'react';
import './style.scss';
import Movie from '../../entity/Movie';
import menuIcon from '../../common/assets/svg/menu-icon.svg';

type MovieCardProps = {
  movie: Movie;
};

const MovieCard = ({ movie }: MovieCardProps): ReactElement => {
  return (
    <div className="card">
      <div className="movie-cover">
        <img className="movie-img" src={movie.imgPath} alt={movie.title} />
        <img className="menu-icon" src={menuIcon} alt="Menu" />
      </div>
      <div className="title-wrapper">
        <p className="title">{movie.title}</p>
        <p className="year">{movie.year}</p>
      </div>
      <p className="genres">{movie.genres.join(', ')}</p>
    </div>
  );
};

export default MovieCard;
