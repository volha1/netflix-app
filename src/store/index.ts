import { configureStore } from '@reduxjs/toolkit';
import movieReducer from './moviesSlice';

const store = configureStore({ reducer: { moviesReducer: movieReducer } });

type AppDispatch = typeof store.dispatch;

export { store, AppDispatch };
