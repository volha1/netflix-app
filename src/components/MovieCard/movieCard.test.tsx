import React from 'react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import * as reduxHooks from 'react-redux';
import { render, screen } from '@testing-library/react';
import * as actions from '../../store/moviesSlice';

import MovieCard from './MovieCard';

jest.mock('react-redux');
const mockedDispatch = jest.spyOn(reduxHooks, 'useDispatch');

describe('MovieCard test', () => {
  const movie = {
    id: '1',
    title: 'title',
    voteAverage: '10',
    releaseDate: '2022',
    imgPath: 'https://test.com',
    genres: ['Horror'],
    overview: 'overview text',
    runtime: 110,
  };
  const onClick = jest.fn();
  const dispatch = jest.fn();
  mockedDispatch.mockReturnValue(dispatch);

  it('MovieCard snapshot', () => {
    const component = renderer.create(<MovieCard movie={movie} onClick={onClick} />);
    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('onClick function works', async () => {
    render(<MovieCard movie={movie} onClick={onClick} />);

    await userEvent.click(screen.getByTestId('card'));

    expect(onClick).toHaveBeenCalledWith({ movie: movie.id });
  });

  it('dispatch for "markMovieForDeletion" works', async () => {
    const mockedMarkMovieForDeletion = jest.spyOn(actions, 'markMovieForDeletion');

    render(<MovieCard movie={movie} onClick={onClick} />);

    await userEvent.click(screen.getByTestId('delete-btn'));

    expect(dispatch).toHaveBeenCalled();
    expect(mockedMarkMovieForDeletion).toHaveBeenCalledWith(movie.id);
  });

  it('dispatch for "saveMovieForEditing" works', async () => {
    const mockedSaveMovieForEditing = jest.spyOn(actions, 'saveMovieForEditing');

    render(<MovieCard movie={movie} onClick={onClick} />);

    await userEvent.click(screen.getByTestId('edit-btn'));

    expect(dispatch).toHaveBeenCalled();
    expect(mockedSaveMovieForEditing).toHaveBeenCalledWith(movie);
  });
});
