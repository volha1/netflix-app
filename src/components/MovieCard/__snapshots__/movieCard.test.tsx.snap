// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard test MovieCard snapshot 1`] = `
<div
  className="card"
  data-testid="card"
  onClick={[Function]}
>
  <div
    className="movie-cover"
  >
    <img
      alt="title"
      className="movie-img"
      onError={[Function]}
      src="https://test.com"
    />
    <img
      alt="Menu"
      className="menu-icon"
      onClick={[Function]}
    />
    <div
      className="menu"
      onClick={[Function]}
    >
      <div
        className="close-btn"
        data-testid="close-btn"
        onClick={[Function]}
      />
      <ul>
        <li>
          <button
            data-testid="edit-btn"
            onClick={[Function]}
            type="button"
          >
            Edit
          </button>
        </li>
        <li>
          <button
            data-testid="delete-btn"
            onClick={[Function]}
            type="button"
          >
            Delete
          </button>
        </li>
      </ul>
    </div>
  </div>
  <div
    className="title-wrapper"
  >
    <p
      className="card-title"
    >
      title
    </p>
    <p
      className="year"
    >
      2022
    </p>
  </div>
  <p
    className="genres"
  >
    Horror
  </p>
</div>
`;
